ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml

```diff
@@ -537,17 +537,30 @@ let monitor_write_loop writers =
       Xenctrl.with_intf (fun xc ->
           while true do
             try
+              let last_loop_start_time = Unix.gettimeofday () in
```

Comment:
I'd prefer if an `Mtime.counter` was used:
```ocaml
let from_iteration_start = Mtime_clock.counter () in

...

let iteration_elapsed = Mtime_clock.count from_iteration_start in
if Mtime.Span.is_longer ~than:!Rrdd_shared.timeslice iteration_elapsed then
  warn ...
else
  let delay =
    Mtime.Span.diff !Rrdd_shared.timeslice iteration_elapsed
    |> Clock.Timer.span_to_s
  in
  Thread.delay delay
```

Reply:
`next_reading` also needs to be updated to use Mtime, it all currently relies on `gettimeofday` which is prone to NTP jumps.

Reply:
Counters probably wouldn't work well, because we need to calculate 2 different spans: begin to end, and current to last end, where end is a single point in time. That one should probably be an Mtime.t then, and we can do arithmetic on Mtime.t (converting to spans if needed, the APIs aren't very pleasant to work with in this regard).

I think Mtime.t should represent a monotonic point in time, which is better than gettimeofday, but still "absolute" in some sense (well it'll be relative to machine start or process start, but not relative to other calls in the code)

Reply:
> Counters probably wouldn't work well, because we need to calculate 2 different spans: begin to end, and current to last end, where end is a single point in time. That one should probably be an Mtime.t then

2 spans should be enough: the amount of time between the start and the last end, and the amount of time between the start and the current end. These can be subtracted to know how much time has passed between them